src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.netstandard.cs

```diff
@@ -36,6 +42,48 @@ internal AssemblyName GetAssemblyName(StringHandle nameHandle, Version version,
             return assemblyName;
         }
 
+        internal static unsafe AssemblyName GetAssemblyName(string assemblyFile)
+        {
+            FileStream? fileStream = null;
+            MemoryMappedFile? mappedFile = null;
+            MemoryMappedViewAccessor? accessor = null;
+            PEReader? peReader = null;
+            try
+            {
+                // Create stream because CreateFromFile(string, ...) uses FileShare.None which is too strict
+                fileStream = new FileStream(assemblyFile, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, false);
+                if (fileStream.Length == 0)
+                {
+                    throw new BadImageFormatException(SR.PEImageDoesNotHaveMetadata, assemblyFile);
+                }
+
+                mappedFile = MemoryMappedFile.CreateFromFile(
+                    fileStream, null, fileStream.Length, MemoryMappedFileAccess.Read, HandleInheritability.None, true);
+                accessor = mappedFile.CreateViewAccessor(0, 0, MemoryMappedFileAccess.Read);
+
+                SafeMemoryMappedViewHandle? safeBuffer = accessor.SafeMemoryMappedViewHandle;
+                peReader = new PEReader((byte*)safeBuffer.DangerousGetHandle(), (int)safeBuffer.ByteLength);
+                MetadataReader mdReader = peReader.GetMetadataReader(MetadataReaderOptions.None);
+                AssemblyName assemblyName = mdReader.GetAssemblyDefinition().GetAssemblyName();
+                return assemblyName;
+            }
+            catch (InvalidOperationException ex)
```

Comment:
Should the catch be after the Dispose so that the file is closed while the exception is thrown? (Matches the behavior of the current unmanaged impl.)

Reply:
I think the end result will be the same, but it does seem cleaner to dispose first.

I had more concerns about what happens if one of the `Dispose` may throw, but we use this pattern in few more places, so I assume it is ok.